tiledb/sm/tile/chunk_buffers.cc

```diff
@@ -0,0 +1,429 @@
+  return buffers_.size();
+}
+
+bool ChunkBuffers::contigious() const {
+  return contigious_;
+}
+
+Status ChunkBuffers::init_fixed_size(
+    const bool contigious,
+    const uint64_t total_size,
+    const uint32_t chunk_size) {
+  if (!buffers_.empty()) {
+    return LOG_STATUS(Status::TileError(
+        "Cannot init chunk buffers; Chunk buffers non-empty."));
+  }
+
+  contigious_ = contigious;
+  chunk_size_ = chunk_size;
+
+  // Calculate the last chunk size.
+  last_chunk_size_ = total_size % chunk_size_;
+  if (last_chunk_size_ == 0) {
+    last_chunk_size_ = chunk_size_;
+  }
+
+  // Calculate the number of chunks required.
+  const size_t nchunks = last_chunk_size_ == chunk_size_ ?
+                             total_size / chunk_size_ :
+                             total_size / chunk_size_ + 1;
+
+  buffers_.resize(nchunks);
+  cached_size_ = chunk_size_ * (buffers_.size() - 1) + last_chunk_size_;
+
+  return Status::Ok();
+}
+
+Status ChunkBuffers::init_var_size(
+    const bool contigious, std::vector<uint32_t>&& var_chunk_sizes) {
+  if (!buffers_.empty()) {
+    return LOG_STATUS(Status::TileError(
+        "Cannot init chunk buffers; Chunk buffers non-empty."));
+  }
+
+  contigious_ = contigious;
+  var_chunk_sizes_ = std::move(var_chunk_sizes);
+  buffers_.resize(var_chunk_sizes_.size());
+
+  assert(cached_size_ == 0);
+  for (const auto& var_chunk_size : var_chunk_sizes_) {
+    cached_size_ += var_chunk_size;
+  }
+
+  return Status::Ok();
+}
+
+Status ChunkBuffers::alloc_discrete(
+    const size_t chunk_idx, void** const buffer) {
+  if (contigious_) {
+    return LOG_STATUS(
+        Status::TileError("Cannot alloc discrete internal chunk buffer; Chunk "
+                          "buffers are contigiously allocated"));
+  }
+
+  if (chunk_idx >= buffers_.size()) {
+    return LOG_STATUS(Status::TileError(
+        "Cannot alloc internal chunk buffer; Chunk index out of bounds"));
+  }
+
+  buffers_[chunk_idx] = malloc(get_chunk_size(chunk_idx));
+  if (!buffers_[chunk_idx]) {
+    LOG_FATAL("malloc() failed");
+  }
+
+  if (buffer) {
+    *buffer = buffers_[chunk_idx];
+  }
+
+  return Status::Ok();
+}
+
+Status ChunkBuffers::free_discrete(const size_t chunk_idx) {
+  if (contigious_) {
+    return LOG_STATUS(
+        Status::TileError("Cannot free discrete internal chunk buffer; Chunk "
+                          "buffers are contigiously allocated"));
+  }
+
+  if (chunk_idx >= buffers_.size()) {
+    return LOG_STATUS(Status::TileError(
+        "Cannot free internal chunk buffer; Chunk index out of bounds"));
+  }
+
+  ::free(buffers_[chunk_idx]);
+  return Status::Ok();
+}
+
+Status ChunkBuffers::set_contigious(void* const buffer) {
+  if (!contigious_) {
+    return LOG_STATUS(
+        Status::TileError("Cannot alloc discrete internal chunk buffer; Chunk "
+                          "buffers are discretely allocated"));
+  }
+
+  if (buffers_.empty()) {
+    return LOG_STATUS(Status::TileError(
```

Comment:
Let's introduce a new `Status::ChunkBuffers` error and replace all `Status::TileError`.